what4/src/What4/Serialize/Log.hs

```diff
@@ -0,0 +1,450 @@
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE NondecreasingIndentation #-}
+-- | Description: Log msgs via a synchronized channel
+--
+-- Log msgs via a synchronized channel.
+--
+-- With inspiration from the @monad-logger@ package.
+--
+-- See examples in 'SemMC.Log.Tests'.
+--
+-- WARNING: loggers that automatically infer the call stack (via
+-- `Ghc.HasCallStack`) are not composable, in that they infer a call
+-- stack at their call site. So, if you use one to build up another
+-- log function, then that derived log function will infer bogus call
+-- sites! Of course, it's pretty easy to write
+--
+--     writeLogEvent logCfg level msg
+--
+-- when defining a new logger, so not a big deal, just something to
+-- watch out for.
+module What4.Serialize.Log (
+  -- * Misc
+  LogLevel(..),
+  LogEvent(..),
+  LogMsg,
+  Ghc.HasCallStack,
+  -- * Implicit param logger interface
+  HasLogCfg,
+  logIO,
+  logTrace,
+  withLogCfg,
+  getLogCfg,
+  -- * Explicit parameter logger interface
+  logIOWith,
+  logEndWith,
+  writeLogEvent,
+  -- * Monadic logger interface
+  MonadHasLogCfg(..),
+  logM,
+  -- * Configuration
+  LogCfg,
+  mkLogCfg,
+  mkNonLogCfg,
+  withLogging,
+  -- * Log consumers
+  stdErrLogEventConsumer,
+  fileLogEventConsumer,
+  tmpFileLogEventConsumer,
+  -- * Log formatting and consumption (useful for 3rd-party consumers)
+  prettyLogEvent,
+  consumeUntilEnd,
+  -- * Named threads
+  named,
+  namedIO,
+  namedM
+  ) where
+
+import qualified GHC.Stack as Ghc
+
+import qualified Control.Concurrent as Cc
+import qualified Control.Exception as Cc
+import           Control.Monad (when)
+
+import qualified Data.Time.Clock as T
+import qualified Data.Time.Format as T
+
+import qualified System.IO as IO
+import qualified System.IO.Unsafe as IO
+
+import qualified UnliftIO as U
+
+import qualified Control.Concurrent.STM as Stm
+import qualified Control.Concurrent.BoundedChan as BC
+import           Control.Monad.IO.Class ( MonadIO, liftIO )
+import           Data.Map.Strict ( Map )
+import qualified Data.Map.Strict as Map
+import           System.Directory ( createDirectoryIfMissing, getTemporaryDirectory )
+import           Text.Printf ( printf )
+
+import Debug.Trace
+
+----------------------------------------------------------------
+-- * API
+
+-- | Log levels, in increasing severity/precedence order.
+data LogLevel = Debug -- ^ Fine details
+              | Info  -- ^ Tracking progress
+              | Warn  -- ^ Something notable or suspicious
+              | Error -- ^ Something bad
+              deriving (Show, Eq, Ord, Read)
+
+type LogMsg = String
+
+-- | Constraints for functions that call the logger.
+--
+-- WARNING: Due to GHC bug, this can't be used:
+-- https://ghc.haskell.org/trac/ghc/ticket/14218
```

Comment:
This bug has since been fixed. Is the `type LogC = (Ghc.HasCallStack, HasLogCfg)` design still worth pursuing? I'm not sure after looking at this module, since the approach that it takes is rather different from that of GHC's `HasCallStack`.

If the design is worth pursuing, it would be worth opening an issue about this. If not, we should just delete this comment.

Reply:
Removed